### PR TITLE
[SPARK-15381][SQL] physical object operator should define reference correctly

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

Whole Stage Codegen depends on `SparkPlan.reference` to do some optimization. For physical object operators, they should be consistent with their logical version and set the `reference` correctly.
## How was this patch tested?

new test in DatasetSuite

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala [hunk 1]
```diff
@@ -28,17 +28,41 @@ import org.apache.spark.sql.catalyst.expressions.objects.Invoke
...
+
+  def inputObjectType: DataType = child.output.head.dataType
+}
+
 /**
  * Takes the input row from child and turns it into object using the given deserializer expression.
  * The output of this operator is a single-field safe row containing the deserialized object.
  */
-case class DeserializeToObject(
+case class DeserializeToObjectExec(
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala]: not related, but to make the name consistent with other physical plans.

**Comment by cloud-fan** [comment]
cc @davies @rxin @yhuai

**Comment by cloud-fan** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala [hunk 2]
```diff
@@ -70,7 +94,7 @@ case class DeserializeToObject(
  */
 case class SerializeFromObjectExec(
     serializer: Seq[NamedExpression],
-    child: SparkPlan) extends UnaryExecNode with CodegenSupport {
+    child: SparkPlan) extends UnaryExecNode with ObjectConsumerExec with CodegenSupport {
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala]: minor: ObjectConsumerExec already extend UnaryExecNode

File: sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala [hunk 3]
```diff
@@ -166,10 +187,7 @@ case class MapElementsExec(
     func: AnyRef,
     outputObjAttr: Attribute,
     child: SparkPlan)
-  extends UnaryExecNode with ObjectOperator with CodegenSupport {
-
-  override def output: Seq[Attribute] = outputObjAttr :: Nil
-  override def producedAttributes: AttributeSet = AttributeSet(outputObjAttr)
+  extends UnaryExecNode with ObjectProducerExec with ObjectConsumerExec with CodegenSupport {
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala]: same here

File: sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala [hunk 4]
```diff
@@ -141,15 +165,12 @@ case class MapPartitionsExec(
     func: Iterator[Any] => Iterator[Any],
     outputObjAttr: Attribute,
     child: SparkPlan)
-  extends UnaryExecNode with ObjectOperator {
-
-  override def output: Seq[Attribute] = outputObjAttr :: Nil
-  override def producedAttributes: AttributeSet = AttributeSet(outputObjAttr)
+  extends UnaryExecNode with ObjectProducerExec with ObjectConsumerExec {
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala]: same here

**Comment by davies** [comment]
LGTM

**Comment by davies** [comment]
Merging this into master and 2.0, thanks!

**Pull request closed without merging**